openedx/core/djangoapps/content/block_structure/models.py

```diff
@@ -0,0 +1,186 @@
+        unicode(data_usage_key),
+    )
+
+
+def _path_name(bs_model, filename):  # pylint:disable=unused-argument
+    """
+    Returns path name to use for the given
+    BlockStructureModel instance.
+    """
+    filename = '{}-{}'.format(
+        unicode(bs_model.data_version),
+        unicode(bs_model.data_edit_timestamp),
+    )
+    return _create_path(
+        _directory_name(bs_model.data_usage_key),
+        filename,
+    )
+
+
+def _bs_model_storage():
+    """
+    Get django Storage object for BlockStructureModel.
+    """
+    return get_storage(
+        settings.BLOCK_STRUCTURES_SETTINGS.get('STORAGE_CLASS'),
+        **settings.BLOCK_STRUCTURES_SETTINGS.get('STORAGE_KWARGS', {})
+    )
+
+
+class BlockStructureModel(TimeStampedModel):
+    """
+    Model for storing Block Structure information.
+    """
+    class Meta(object):
+        db_table = 'block_structure'
+
+    data_usage_key = UsageKeyField(
+        u'Identifier of the data being collected.',
+        blank=False,
+        max_length=255,
+        unique=True,
+    )
+    data_version = models.CharField(
+        u'Version of the data at the time of collection.',
+        blank=True,
+        max_length=255,
+    )
+    data_edit_timestamp = models.DateTimeField(
+        u'Edit timestamp of the data at the time of collection.',
+        blank=True,
+        null=True,
+    )
+    transformers_schema_version = models.CharField(
+        u'Representation of the schema version of the transformers used during collection.',
+        blank=False,
+        max_length=255,
+    )
+    block_structure_schema_version = models.CharField(
+        u'Version of the block structure schema at the time of collection.',
+        blank=False,
+        max_length=255,
+    )
+    data = models.FileField(
+        upload_to=_path_name,
+        storage=_bs_model_storage(),
+    )
+
+    def get_serialized_data(self):
+        """
+        Returns the collected data for this instance.
+        """
+        return self.data.read()
+
+    @classmethod
+    def get(cls, data_usage_key):
+        """
+        Returns the entry associated with the given data_usage_key.
+        Raises:
+             BlockStructureNotFound if an entry for data_usage_key is not found.
+        """
+        try:
+            return cls.objects.get(data_usage_key=data_usage_key)
+        except cls.DoesNotExist:
+            raise BlockStructureNotFound(data_usage_key)
+
+    @classmethod
+    def update_or_create(cls, serialized_data, data_usage_key, **kwargs):
+        """
+        Updates or creates the BlockStructureModel entry
+        for the given data_usage_key in the kwargs,
+        uploading serialized_data as the content data.
+        """
+        bs_model, created = cls.objects.update_or_create(defaults=kwargs, data_usage_key=data_usage_key)
+        bs_model.data.save('', ContentFile(serialized_data))
+        if not created:
```

Comment:
Why `if not created` here? If this is still designed to be append only, `created` will always be true (b/c we're never updating anything), so we'll never delete old files.